src/main.py

```diff
@@ -0,0 +1,119 @@
+        
+        # Check if a supported version of memtester is installed in the system
+        def version_callback(version, is_known):
+            step.add_log(tv.LogSeverity.INFO, "Memtester v{} was found".format(version))
+            if (not is_known):
+                m = "This version of memtester was not tested. Expect parsing errors"
+                step.add_log(tv.LogSeverity.WARNING, m)
+        observer.callbacks.version_ready = version_callback
+        
+        # Report loop results
+        def loop_callback(loop):
+            tests = loop.failed_tests()
+            if (len(tests) == 0):
+                m = "Loop #{} finished with success".format(loop.index)
+                step.add_log(tv.LogSeverity.INFO, m)
+            else:
+                # Log failed tests
+                names = ", ".join([t.name for t in tests])
+                m = "Loop #{} failed {} tests: {}".format(loop.index, len(tests), names)
+                step.add_error(symptom="loop-failure", message=m)
+                
+                # Record failed addresses
+                for test in tests:
+                    for r in test.result:
+                        bad_addrs.add_measurement(value=r.addr)
+        observer.callbacks.loop_ready = loop_callback
+        
+        # Report individual tests to make long runs more responsive
+        def test_callback(test):
+            if test.passed():
+                step.add_log(tv.LogSeverity.INFO, "Test '{}' passed".format(test.name))
+            else:
+                step.add_log(tv.LogSeverity.ERROR, "Test '{}' failed".format(test.name))
+        observer.callbacks.test_ready = test_callback
+        
+        # Report diagnosis
+        def run_callback(failed_loop_count):
+            if failed_loop_count == 0:
+                step.add_diagnosis(tv.DiagnosisType.PASS, verdict="memtester-passed")
+            else:
+                step.add_diagnosis(tv.DiagnosisType.FAIL, verdict="memtester-failed")
+                raise tv.TestRunError(status=tv.TestStatus.COMPLETE, result=tv.TestResult.FAIL)
+        observer.callbacks.run_ready = run_callback
+        
+        # Report parsing errors (supposed to only happen with unknown memtester versions)
+        def parsing_error_callback(desc):
+            step.add_error(symptom="memtester-parsing-error", message=desc)
+            raise tv.TestRunError(status=tv.TestStatus.ERROR, result=tv.TestResult.NOT_APPLICABLE)
+        observer.callbacks.parsing_error = parsing_error_callback
+
+        # Log raw memtester output if necessary
+        if args.mt_log_filename:
+            estack.enter_context(log_file := open(args.mt_log_filename, "w"))
+            def line_ready_callback(line):
+                log_file.write(line)
+            observer.callbacks.line_ready = line_ready_callback
+        
+        # Run memtester (finally!)
+        try:
+            mt_cmd = sh.Command(args.mt_path)
+            mt_args = args.mt_args.split(" ")
```

Comment:
does memtester have any args that contain spaces (either in key or value)? if so, this should split incorrectly, eg. `--mt_args "--some-param \"v1 v2\""`

Reply:
>does memtester have any args that contain spaces?

If I understand the usage hint correctly, it does not:

>Usage: memtester [-p physaddrbase [-d device]] <mem>[B|K|M|G] [loops]

But it is better to fix it anyway. Done.